repos/system_upgrade/el7toel8/libraries/fetch.py

```diff
@@ -0,0 +1,73 @@
+import os
+
+import requests
+
+from leapp.exceptions import StopActorExecutionError
+from leapp.libraries.common.config import get_env
+from leapp.libraries.stdlib import api
+
+SERVICE_HOST_DEFAULT = "https://cert.cloud.redhat.com"
+PROXY_HOST_DEFAULT = "http://squid.corp.redhat.com:3128"
+REQUEST_TIMEOUT = 5
+
+
+def _raise_error(local_path, details):
+    """
+    If the file acquisition fails in any way, throw an informative error to stop the actor.
+    """
+    summary = "Data file {lp} is invalid or could not be retrieved.".format(lp=local_path)
+    hint = ("Read documentation at: https://access.redhat.com/articles/3664871"
+            " for more information about how to retrieve the file.")
+
+    raise StopActorExecutionError(summary, details={'details': details, 'hint': hint})
+
+
+def read_or_fetch(filename, directory="/etc/leapp/files", service=None, allow_empty=False):
+    """
+    Return contents of a file or fetch them from an online service if the file does not exist.
+    """
+    logger = api.current_logger()
+    local_path = os.path.join(directory, filename)
+
+    # try to get the data locally
+    if not os.path.exists(local_path):
+        logger.warning("File {lp} does not exist, falling back to online service".format(lp=local_path))
+    else:
+        try:
+            with open(local_path) as f:
+                data = f.read()
+                if not allow_empty and not data:
+                    _raise_error(local_path, "File {lp} exists but is empty".format(lp=local_path))
+                logger.warning("File {lp} successfully read ({l} bytes)".format(lp=local_path, l=len(data)))
+                return data
+        except EnvironmentError:
+            _raise_error(local_path, "File {lp} exists but couldn't be read".format(lp=local_path))
+        except Exception as e:
+            raise e
+
+    # if the data is not present locally, fetch it from the online service
+    service = service or get_env("LEAPP_SERVICE_HOST", default=SERVICE_HOST_DEFAULT)
+    service_path = "{s}/api/pes/{f}".format(s=service, f=filename)
+    proxy = PROXY_HOST_DEFAULT or get_env("LEAPP_PROXY_HOST", default=PROXY_HOST_DEFAULT)
+    proxies = {"https": proxy} if proxy else None
+    cert = ("/etc/pki/consumer/cert.pem", "/etc/pki/consumer/key.pem")
```

Comment:
The path `/etc/pki/consumer` should be obtained from `consumerCertDir` variable in `/etc/rhsm/rhsm.conf`

Reply:
@shaded-enmity  do you expect this to be +/- "common" case? If we expect different path just in rare cases, I would go with the fix in the next release regarding the time.

Reply:
@pirat89 that's what `insights-client` is doing, I'd assume we can just get *inspired* by that code. But worse case scenario you're right, we can fix this one later in which case I'll need some tracking ticket for that.

Reply:
@shaded-enmity  no need. I will cover it in refactoring. in case it will not be fixed here, you can put the fix into the AC (I have a ticket for the refactoring already). As well, if this change happens, it should be covered by tests @mmacura311